File: day15/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <sstream>
#include  <regex>
#include <optional>
#include <functional>
#include <set>
#include <unordered_map>
#include <map>
#include <list>

int64_t hash(const std::string& line) {
    int64_t result = 0;
    for (char c: line) {
        result += static_cast<int64_t>(c);
        result *= 17;
        result %= 256;
    }
    return result;
}

void part1(const std::vector<std::string>& lines) {
    int64_t result = 0;
    for (const auto& line: lines) {
        result += hash(line);
    }

    std::cout << "part1: " << result << std::endl;
}

void part2(const std::vector<std::string>& lines) {
    std::vector<std::vector<std::pair<std::string, int64_t>>> boxes(256);
    for (const auto& line: lines) {
        auto pos = line.find('-');
        if (pos != std::string::npos) {
            auto label = line.substr(0, pos);
            std::erase_if(boxes[hash(label)], [&label](const auto& elem) { return elem.first == label; });
        } else {
            auto pos = line.find('=');
            auto label = line.substr(0, pos);
            int64_t focal = std::stoll(line.substr(pos + 1));
            auto& box = boxes[hash(label)];
            auto it = std::find_if(box.begin(), box.end(), [&label](const auto& elem) { return elem.first == label; });
            if (it == box.end()) {
                box.push_back({label, focal});
            } else {
                *it = {label, focal};
            }
        }
    }

    int64_t result = 0;
    for (int64_t box_idx = 0; box_idx < boxes.size(); ++box_idx) {
        const auto& box = boxes[box_idx];
        for (int64_t slot_idx = 0; slot_idx < box.size(); ++slot_idx) {
            int64_t power = (box_idx + 1) * (slot_idx + 1) * box[slot_idx].second;
            result += power;
        }
    }

    std::cout << "part2: " << result << std::endl;
}

int main() {
    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");

    std::string line;
    std::vector<std::string> lines;
    while (std::getline(ifs, line, ',')) {
        std::erase_if(line, [](char c) { return c == '\n'; });
        lines.push_back(line);
    }

    part1(lines);
    part2(lines);

    return 0;
}

File: day15/src/main.rs
use gcd::binary_u64;
use lazy_static::lazy_static;
use regex::{Match, Regex};
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;
use std::vec;

fn read_file<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let Ok(lines) = read_file("input.txt") else {
        panic!("file is not found");
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn hash(input: &str) -> usize {
    let mut result = 0usize;
    for ch in input.chars() {
        result += ch as usize;
        result *= 17;
        result %= 256;
    }
    result
}

fn main() {
    let line = read().join("");
    let table = line.split(",").collect::<Vec<_>>();

    println!("part1 {:?}", table.iter().map(|&x| hash(x)).sum::<usize>());

    let mut boxes: Vec<Vec<(&str, i64)>> = vec![vec![]; 256];

    for elem in table {
        if let Some(pos) = elem.find(|x| x == '-') {
            let key = &elem[0..pos];
            let box_idx = hash(key);
            if let Some(idx) = boxes[box_idx].iter().position(|&(x, _)| x == key) {
                boxes[box_idx].remove(idx);
            }
        } else if let Some(pos) = elem.find(|x| x == '=') {
            let key = &elem[0..pos];
            let focal = elem[pos + 1..].parse::<i64>().unwrap();
            let box_idx = hash(key);
            if let Some(idx) = boxes[box_idx].iter().position(|&(x, _)| x == key) {
                boxes[box_idx][idx].1 = focal;
            } else {
                boxes[box_idx].push((key, focal));
            }
        }
    }

    let mut result = 0i64;
    for (box_idx, boxx) in boxes.iter().enumerate() {
        for (slot_idx, &(_, focal)) in boxx.iter().enumerate() {
            result += (box_idx + 1) as i64 * (slot_idx + 1) as i64 * focal;
        }
    }

    println!("part2 {:?}", result);
}
